addons/binding/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/thirdgeneration/ThirdGenerationHandler.java

```diff
@@ -0,0 +1,493 @@
+     * @param inverterType the type of the device
+     */
+    public ThirdGenerationHandler(Thing thing, HttpClient httpClient, ThirdGenerationInverterTypes inverterType) {
+        super(thing);
+        this.inverterType = inverterType;
+        this.httpClient = httpClient;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // All channels are readonly and updated by the scheduler
+    }
+
+    @Override
+    public void dispose() {
+        if (refreshScheduler != null) {
+            refreshScheduler.cancel(true);
+			refreshScheduler = null;
+        }
+        super.dispose();
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(ThirdGenerationConfiguration.class);
+        // temporary value while initializing
+        updateStatus(ThingStatus.UNKNOWN);
+
+        // Start the authentication
+        scheduler.schedule(this::authenticate, 1, TimeUnit.SECONDS);
+
+        // Start the update scheduler as configured
+        refreshScheduler = scheduler.scheduleWithFixedDelay(this::updateChannelValues, 10,
+                config.refreshInternalInSeconds, TimeUnit.SECONDS);
+    }
+
+    /**
+     * The API supports the resolution of multiple values at a time
+     *
+     * Therefore this methods builds one request to gather all information for the current inverter.
+     * The list contains all channels as defined in {@link ThirdGenerationMappingInverterToChannel} for the
+     * current inverter
+     *
+     */
+    void updateChannelValues() {
+        Map<String, List<ThirdGenerationChannelMappingToWebApi>> channelList = ThirdGenerationMappingInverterToChannel
+                .getModuleToChannelsMappingForInverter(inverterType);
+        JsonArray updateMessageJsonArray = getUpdateChannelMessage(channelList);
+
+        // Send the API request to get values for all channels
+        ContentResponse updateMessageContentResponse;
+        try {
+            updateMessageContentResponse = ThirdGenerationHttpHelper.executeHttpPost(httpClient, config.url,
+                    PROCESSDATA, updateMessageJsonArray, sessionId);
+            if (updateMessageContentResponse.getStatus() == 404) {
+                // Module not found
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                        COMMUNICATION_ERROR_INCOMPATIBLE_DEVICE);
+                return;
+            }
+            if (updateMessageContentResponse.getStatus() == 503) {
+                // Communication error (e.g. during initial boot of the SCB)
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                        COMMUNICATION_ERROR_HTTP);
+                return;
+            }
+            if (updateMessageContentResponse.getStatus() == 401) {
+                // session not valid (timed out?)
+                logger.info("Session expired");
```

Comment:
Since this is logged to info it might be better to make the message a little more descriptive. For example add something about it's going to retry authenticating so the user knows what is happening.